File: src/dreambot_cpp_examples/src/simple_publisher.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>

#include <chrono>

using namespace std::chrono_literals;

class SimplePublisher : public rclcpp::Node {
public:
    SimplePublisher() 
    : Node("simple_publisher"), counter_(0) {
        publisher_ = create_publisher<std_msgs::msg::String>("chatter", 10);
        timer_ = create_wall_timer(
            1s, std::bind(&SimplePublisher::publishMessage, this)
        );
        RCLCPP_INFO(get_logger(), "Publisher initialized at 1 Hz");
    }

private:
    void publishMessage() {
        std_msgs::msg::String message;
        message.data = "Hello ROS 2 : " + std::to_string(counter_++);
        publisher_->publish(message);
    }

    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    unsigned int counter_;
};

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SimplePublisher>());
    rclcpp::shutdown();
    return 0;
}

File: src/dreambot_py_examples/dreambot_py_examples/simple_publisher.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String

class SimplePublisher(Node):
    def __init__(self):
        super().__init__("simple_publisher")
        self.pub_ = self.create_publisher(String, "chatter", 10)
        self.counter_ = 0
        self.frequency_ = 1.0  # Hz
        self.get_logger().info("Publishing at %.1f Hz" % self.frequency_)
        self.timer_ = self.create_timer(1.0 / self.frequency_, self.timer_callback)

    def timer_callback(self):
        msg = String()
        msg.data = "Hello ROS 2 - counter: %d" % self.counter_
        self.get_logger().info(msg.data)
        self.pub_.publish(msg)
        self.counter_ += 1

def main():
    rclpy.init()
    node = SimplePublisher()  # ✅ Avoid reusing class name
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()
